Complete `EraScript` hierarchy with missing  classes

teodanciu opened the pull request:
# Description

<!-- Add your description here, if it fixes a particular issue please provide a
[link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=)
to the issue. -->

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

lehins commented on eras/allegra/impl/src/Cardano/Ledger/Allegra/Scripts.hs:
```diff
@@ -130,6 +146,13 @@ data TimelockRaw era
 
+class ShelleyEraScript era => AllegraEraScript era where
+  mkTimeStart :: SlotNo -> NativeScript era
+  getTimeStart :: NativeScript era -> Maybe (SlotNo)
```
```suggestion
  getTimeStart :: NativeScript era -> Maybe SlotNo
```

lehins commented on eras/allegra/impl/src/Cardano/Ledger/Allegra/Scripts.hs:
```diff
@@ -130,6 +146,13 @@ data TimelockRaw era
+  getTimeStart :: NativeScript era -> Maybe (SlotNo)
+
+  mkTimeExpire :: SlotNo -> NativeScript era
+  getTimeExpire :: NativeScript era -> Maybe (SlotNo)
```
```suggestion
  getTimeExpire :: NativeScript era -> Maybe SlotNo
```

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Scripts.hs:
```diff
@@ -13,15 +13,17 @@
-    RequireSignature,
-    RequireMOf
-  ),
+  MultiSig (..),
```
You can't export raw constructors, because that will expose the `MultiSigConstr`
```suggestion
  MultiSig,
```

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Scripts.hs:
```diff
@@ -124,6 +142,29 @@ instance Crypto c => EraScript (ShelleyEra c) where
+
+  mkRequireAllOf ms =
+    MultiSigConstr $ memoBytes (Sum RequireAllOf' 1 !> To (toList ms))
+  getRequireAllOf (MultiSigConstr (Memo (RequireAllOf' ms) _)) = Just (StrictSeq.fromList ms)
```
I think it might be better to change the actual type in MultiSig to use StrictSeq instead of list
```suggestion
  getRequireAllOf (MultiSigConstr (Memo (RequireAllOf' ms) _)) = Just ms
```

teodanciu replied:
Ah I was afraid it might cause problems, but I guess the serialization would change.

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Scripts.hs:
```diff
@@ -197,25 +230,26 @@ eqMultiSigRaw t1 t2 = go (getMemoRawType t1) (getMemoRawType t2)
 -- key hashes that signed the transaction to be validated.
 evalMultiSig ::
-  Era era =>
+  ShelleyEraScript era =>
```
If you don't do this restriction then all of the sudden the `"Impossible"` case will be possible for timelocks
```suggestion
  (ShelleyEraScript era, NativeScript era ~ MultiSig era) =>
```

teodanciu replied:
Makes sense, thank you

lehins commented on eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/Arbitrary.hs:
```diff
@@ -88,6 +97,8 @@ import Test.Cardano.Ledger.Core.Arbitrary ()
 import Test.Cardano.Ledger.Core.Utils (unsafeBoundRational)
 import Test.QuickCheck.Hedgehog (hedgehog)
 
+import Data.Sequence.Strict (fromList)
```
```suggestion
import Data.Sequence.Strict (fromList)
```

lehins reviewed (commented):
This is looking great!

teodanciu committed:
Introduce `ShelleyEraScript`,`AllegraEraScript` to `EraScript` hierarchy

Replace era-specific patterns with new ones, constrained by the new
EraScript classes, that build and access `NativeScript` instead of
concrete types

teodanciu committed:
Stop exporting `MultiSig` constructor

teodanciu committed:
Change underlying representation of MultiSig to StrictSeq

lehins commented on eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/MultiSigExamples.hs:
```diff
@@ -95,39 +97,48 @@ _assertScriptHashSizeMatchesAddrHashSize (ScriptHash h) =
 -- Multi-signature scripts
-singleKeyOnly :: Era era => Addr (EraCrypto era) -> MultiSig era
+singleKeyOnly ::
+  (ShelleyEraScript era, NativeScript era ~ MultiSig era) =>
```
I don't think any of the `NativeScript era ~ MultiSig era` constraints in this module are necessary
```suggestion
  ShelleyEraScript era =>
```

teodanciu replied:
How odd.. I'm sure I  added them under duress from the compiler rather than arbitrarily, but you're right, they're not needed... Not sure why that happened, maybe an intermediary state or something :( Bah . 
Thanks!

lehins commented on eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/MultiSigExamples.hs:
```diff
@@ -95,39 +97,48 @@ _assertScriptHashSizeMatchesAddrHashSize (ScriptHash h) =
 singleKeyOnly _ = error "use VKey address"
 
-aliceOnly :: Era era => MultiSig era
+aliceOnly :: (ShelleyEraScript era, NativeScript era ~ MultiSig era) => NativeScript era
```
```suggestion
aliceOnly :: ShelleyEraScript era => NativeScript era
```

lehins commented on eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/MultiSigExamples.hs:
```diff
@@ -95,39 +97,48 @@ _assertScriptHashSizeMatchesAddrHashSize (ScriptHash h) =
 aliceOnly = singleKeyOnly Cast.aliceAddr
 
-bobOnly :: Era era => MultiSig era
+bobOnly :: (ShelleyEraScript era, NativeScript era ~ MultiSig era) => NativeScript era
```
```suggestion
bobOnly :: ShelleyEraScript era => NativeScript era
```

lehins commented on eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/MultiSigExamples.hs:
```diff
@@ -95,39 +97,48 @@ _assertScriptHashSizeMatchesAddrHashSize (ScriptHash h) =
 
-aliceOrBob :: Era era => MultiSig era
-aliceOrBob = RequireAnyOf [aliceOnly, singleKeyOnly Cast.bobAddr]
+aliceOrBob :: (ShelleyEraScript era, NativeScript era ~ MultiSig era) => NativeScript era
```
```suggestion
aliceOrBob :: ShelleyEraScript era => NativeScript era
```

lehins commented on eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/MultiSigExamples.hs:
```diff
@@ -95,39 +97,48 @@ _assertScriptHashSizeMatchesAddrHashSize (ScriptHash h) =
 
-aliceAndBob :: Era era => MultiSig era
-aliceAndBob = RequireAllOf [aliceOnly, singleKeyOnly Cast.bobAddr]
+aliceAndBob :: (ShelleyEraScript era, NativeScript era ~ MultiSig era) => NativeScript era
```
```suggestion
aliceAndBob :: ShelleyEraScript era => NativeScript era
```

lehins commented on eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/MultiSigExamples.hs:
```diff
@@ -95,39 +97,48 @@ _assertScriptHashSizeMatchesAddrHashSize (ScriptHash h) =
 
-aliceAndBobOrCarl :: Era era => MultiSig era
-aliceAndBobOrCarl = RequireMOf 1 [aliceAndBob, singleKeyOnly Cast.carlAddr]
+aliceAndBobOrCarl :: (ShelleyEraScript era, NativeScript era ~ MultiSig era) => NativeScript era
```
```suggestion
aliceAndBobOrCarl :: ShelleyEraScript era => NativeScript era
```

lehins commented on eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/MultiSigExamples.hs:
```diff
@@ -95,39 +97,48 @@ _assertScriptHashSizeMatchesAddrHashSize (ScriptHash h) =
-aliceAndBobOrCarlAndDaria :: Era era => MultiSig era
+aliceAndBobOrCarlAndDaria ::
+  (ShelleyEraScript era, NativeScript era ~ MultiSig era) =>
+  NativeScript era
```
```suggestion
aliceAndBobOrCarlAndDaria :: ShelleyEraScript era => NativeScript era
```

lehins commented on eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/MultiSigExamples.hs:
```diff
@@ -95,39 +97,48 @@ _assertScriptHashSizeMatchesAddrHashSize (ScriptHash h) =
-aliceAndBobOrCarlOrDaria :: Era era => MultiSig era
+aliceAndBobOrCarlOrDaria ::
+  (ShelleyEraScript era, NativeScript era ~ MultiSig era) =>
+  NativeScript era
```
```suggestion
aliceAndBobOrCarlOrDaria :: ShelleyEraScript era => NativeScript era
```

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/Properties.hs:
```diff
@@ -81,6 +81,8 @@ import Test.Control.State.Transition.Trace.Generator.QuickCheck (HasTrace (..))
 import Test.QuickCheck
 import Test.Tasty (TestTree, defaultMain, testGroup)
 
```
Avoid having newlines between imports, cause then fourmolu won't sort it properly
```suggestion
```

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/Properties.hs:
```diff
@@ -111,6 +113,16 @@ genTxAndUTXOState proof@Shelley gsize = do
+genTxAndLEDGERStateShelley ::
+  GenSize -> Gen (TRC (EraRule "LEDGER" S.Shelley), GenState S.Shelley)
+genTxAndLEDGERStateShelley genSize = do
+  (Box _ trc genState) <- genTxAndLEDGERState Shelley genSize
```
Redundant parenthesis 
```suggestion
  Box _ trc genState <- genTxAndLEDGERState Shelley genSize
```

lehins commented on eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/Fees.hs:
```diff
@@ -424,14 +429,16 @@ txWithMDBytes16 =
 
 -- | Spending from a multi-sig address
-msig :: forall era. Era era => MultiSig era
+msig :: forall era. (ShelleyEraScript era, NativeScript era ~ MultiSig era) => MultiSig era
```
Redundant constraint
```suggestion
msig :: ShelleyEraScript era => NativeScript era
```

lehins commented on eras/shelley-ma/test-suite/test/Test/Cardano/Ledger/ShelleyMA/Serialisation/Timelocks.hs:
```diff
@@ -31,13 +41,13 @@ import Test.Tasty.QuickCheck (testProperty)
 -- ================================================================
 
-s1 :: Timelock Shelley
+s1 :: (AllegraEraScript era, NativeScript era ~ Timelock era) => Timelock era
```
No need to specify Timelock anymore
```suggestion
s1 :: AllegraEraScript era => NativeScript era
```

lehins commented on eras/shelley-ma/test-suite/test/Test/Cardano/Ledger/ShelleyMA/Serialisation/Timelocks.hs:
```diff
@@ -31,13 +41,13 @@ import Test.Tasty.QuickCheck (testProperty)
 s1 = RequireAllOf (fromList [RequireTimeStart (SlotNo 12), RequireTimeExpire 18])
 
-s2 :: Timelock Shelley
+s2 :: (AllegraEraScript era, NativeScript era ~ Timelock era) => Timelock era
```
```suggestion
s2 :: AllegraEraScript era => NativeScript era
```

lehins commented on eras/shelley-ma/test-suite/test/Test/Cardano/Ledger/ShelleyMA/Serialisation/Timelocks.hs:
```diff
@@ -31,13 +41,13 @@ import Test.Tasty.QuickCheck (testProperty)
 s2 = RequireAllOf (fromList [RequireTimeStart (SlotNo 12), RequireTimeExpire (SlotNo 23)])
 
-s3 :: Timelock Shelley
+s3 :: (AllegraEraScript era, NativeScript era ~ Timelock era) => Timelock era
```
```suggestion
s3 :: AllegraEraScript era => NativeScript era
```

lehins commented on eras/shelley-ma/test-suite/test/Test/Cardano/Ledger/ShelleyMA/Serialisation/Golden/Encoding.hs:
```diff
@@ -56,7 +69,7 @@ import Test.Tasty.HUnit (Assertion, testCase)
 -- ===============================================
 
-policy1 :: Era era => Timelock era
+policy1 :: (ShelleyEraScript era, NativeScript era ~ Timelock era) => Timelock era
```
```suggestion
policy1 :: ShelleyEraScript era => NativeScript era
```

lehins commented on eras/shelley-ma/test-suite/test/Test/Cardano/Ledger/ShelleyMA/Serialisation/Golden/Encoding.hs:
```diff
@@ -112,7 +125,10 @@ testUpdate =
+scriptGoldenTest ::
+  forall era.
+  (AllegraEraScript era, NativeScript era ~ Timelock era) =>
+  TestTree
```
```suggestion
scriptGoldenTest :: forall era. AllegraEraScript era => TestTree
```

lehins commented on eras/shelley-ma/test-suite/test/Test/Cardano/Ledger/ShelleyMA/Serialisation/Golden/Encoding.hs:
```diff
@@ -170,7 +186,10 @@ metadataNoScriptsGoldenTest =
 
 -- CONTINUE also Scripts
-metadataWithScriptsGoldenTest :: forall era. Era era => TestTree
+metadataWithScriptsGoldenTest ::
+  forall era.
+  (ShelleyEraScript era, NativeScript era ~ Timelock era) =>
+  TestTree
```
```suggestion
metadataWithScriptsGoldenTest :: forall era. ShelleyEraScript era => TestTree
```

lehins commented on eras/shelley-ma/test-suite/src/Test/Cardano/Ledger/MaryEraGen.hs:
```diff
@@ -124,7 +129,10 @@ maryGenesisValue (GenEnv _ _ Constants {minGenesisOutputVal, maxGenesisOutputVal
+trivialPolicy ::
+  (AllegraEraScript era, NativeScript era ~ Timelock era) =>
+  Int ->
+  NativeScript era
```
```suggestion
trivialPolicy :: AllegraEraScript era => Int -> NativeScript era
```

lehins commented on eras/shelley-ma/test-suite/src/Test/Cardano/Ledger/MaryEraGen.hs:
```diff
@@ -142,7 +150,7 @@ coloredCoinMaxMint = 1000 * 1000
 --------------------------------------------------------
 
-redCoins :: Era era => Timelock era
+redCoins :: (AllegraEraScript era, NativeScript era ~ Timelock era) => NativeScript era
```
```suggestion
redCoins :: AllegraEraScript era => NativeScript era
```

lehins commented on eras/shelley-ma/test-suite/src/Test/Cardano/Ledger/MaryEraGen.hs:
```diff
@@ -163,7 +171,7 @@ genRed = do
 --------------------------------------------------------
 
-blueCoins :: Era era => Timelock era
+blueCoins :: (AllegraEraScript era, NativeScript era ~ Timelock era) => NativeScript era
```
```suggestion
blueCoins :: AllegraEraScript era => NativeScript era
```

lehins commented on eras/shelley-ma/test-suite/src/Test/Cardano/Ledger/MaryEraGen.hs:
```diff
@@ -194,7 +202,7 @@ genBlue = do
 --------------------------------------------------------
 
-yellowCoins :: Era era => Timelock era
+yellowCoins :: (AllegraEraScript era, NativeScript era ~ Timelock era) => NativeScript era
```
```suggestion
yellowCoins :: AllegraEraScript era => NativeScript era
```

lehins commented on eras/shelley-ma/test-suite/src/Test/Cardano/Ledger/MaryEraGen.hs:
```diff
@@ -216,7 +224,9 @@ genYellow = do
-policyIndex :: Era era => Map (PolicyID (EraCrypto era)) (Timelock era)
+policyIndex ::
+  (AllegraEraScript era, NativeScript era ~ Timelock era) =>
+  Map (PolicyID (EraCrypto era)) (NativeScript era)
```
```suggestion
policyIndex ::
  AllegraEraScript era => Map (PolicyID (EraCrypto era)) (NativeScript era)
```

lehins commented on eras/shelley-ma/test-suite/src/Test/Cardano/Ledger/MaryEraGen.hs:
```diff
@@ -280,8 +290,10 @@ addTokens _proxy _ _ _ StrictSeq.Empty = Nothing
+  , AllegraEraScript era
   , Value era ~ MaryValue (EraCrypto era)
   , TxOut era ~ ShelleyTxOut era
+  , NativeScript era ~ Timelock era
```
```suggestion
```

lehins commented on eras/shelley-ma/test-suite/src/Test/Cardano/Ledger/AllegraEraGen.hs:
```diff
@@ -201,9 +219,9 @@ partition splits scripts =
 intervals ::
   forall era.
-  Era era =>
+  (AllegraEraScript era, NativeScript era ~ Timelock era) =>
```
```suggestion
  AllegraEraScript era =>
```

lehins commented on eras/shelley-ma/test-suite/src/Test/Cardano/Ledger/AllegraEraGen.hs:
```diff
@@ -180,19 +198,19 @@ genValidityInterval cs@(SlotNo currentSlot) =
 partition ::
   forall era.
-  Era era =>
+  (AllegraEraScript era, NativeScript era ~ Timelock era) =>
```
```suggestion
  AllegraEraScript era =>
```

lehins commented on eras/shelley-ma/test-suite/src/Test/Cardano/Ledger/AllegraEraGen.hs:
```diff
@@ -180,19 +198,19 @@ genValidityInterval cs@(SlotNo currentSlot) =
 someLeaf ::
   forall era.
-  Era era =>
+  (AllegraEraScript era, NativeScript era ~ Timelock era) =>
```
```suggestion
  AllegraEraScript era =>
```

lehins commented on eras/shelley-ma/test-suite/src/Test/Cardano/Ledger/AllegraEraGen.hs:
```diff
@@ -126,13 +138,19 @@ instance CryptoClass.Crypto c => MinGenTxout (AllegraEra c) where
+unQuantifyTL ::
+  (AllegraEraScript era, NativeScript era ~ Timelock era) =>
+  Quantifier (Timelock era) ->
+  Timelock era
```
```suggestion
  AllegraEraScript era => Quantifier (NativeScript era) -> NativeScript era
```

lehins commented on eras/shelley-ma/test-suite/src/Test/Cardano/Ledger/AllegraEraGen.hs:
```diff
@@ -126,13 +138,19 @@ instance CryptoClass.Crypto c => MinGenTxout (AllegraEra c) where
+quantifyTL ::
+  (AllegraEraScript era, NativeScript era ~ Timelock era) =>
+  Timelock era ->
+  Quantifier (Timelock era)
```
```suggestion
quantifyTL :: AllegraEraScript era => NativeScript era -> Quantifier (NativeScript era)
```

lehins commented on eras/shelley-ma/test-suite/src/Test/Cardano/Ledger/Allegra/Examples/Consensus.hs:
```diff
@@ -57,7 +59,7 @@ exampleAllegraTxBody value =
       AuxiliaryDataHash $ mkDummySafeHash (Proxy @(EraCrypto era)) 30
 
-exampleTimelock :: Era era => Timelock era
+exampleTimelock :: (AllegraEraScript era, NativeScript era ~ Timelock era) => Timelock era
```
```suggestion
exampleTimelock :: AllegraEraScript era => NativeScript era
```

lehins reviewed (approved):
There are quite a few places in testing where redundant constraint was specified (eg. `NativeScript era ~ Timelock era`). Such constraints are only really needed for creating class instances or whenever we need to enforce equality, like in `validateMultiSig`

Aside from that it looks great! Thank you!
